### PR TITLE
chore: add some helpers

### PR DESCRIPTION
**Pull request by noahdietz**
This adds a few helpers that will be useful in implementing #972. I started on that issue, and while I didn't get far, I figured these would be useful. @acamadeo will be taking on that issue now, so I want them to have the helpers.

### CODE REVIEW

File: rules/internal/utils/http.go [hunk 1]
```diff
@@ -22,6 +22,13 @@ import (
 	"google.golang.org/protobuf/proto"
 )
 
+// HasHTTPRules returns whether the given method descriptor is annotated with
```
Comment (GregFurmanek) [inline rules/internal/utils/http.go:25]: How about?

HasHTTPRules returns true when ...

File: rules/internal/utils/find_test.go [hunk 2]
```diff
@@ -43,3 +44,46 @@ func TestFindMessage(t *testing.T) {
...
+
+			PublishingInfo publishing_info = 2;
+		}
+	`)
+	msg := file.GetMessageTypes()[0]
+
+	for _, tst := range []struct {
+		name, path string
+	}{
+		{"simple", "parent"},
```
Comment (GregFurmanek) [inline rules/internal/utils/find_test.go:74]: simple -> top_level ?

File: rules/internal/utils/find.go [hunk 3]
```diff
@@ -126,3 +126,26 @@ func GetRepeatedMessageFields(m *desc.MessageDescriptor) []*desc.FieldDescriptor
 
 	return fields
 }
+
+// FindFieldDotNotation attempts to find the field within the given message
+// identified by the dot-notation path e.g. book.name. This is especially useful
+// for resolving path variables in google.api.http and nested
+// fields in google.api.method_signature annotations.
+func FindFieldDotNotation(msg *desc.MessageDescriptor, f string) *desc.FieldDescriptor {
```
Comment (GregFurmanek) [inline rules/internal/utils/find.go:134]: In essence this works like jquery, right?

So the method description might be:

Returns a field descriptor from a passed in message that corresponds to dot separated path. Example: ...

If path not found the method returns nil.

Comment (GregFurmanek) [inline rules/internal/utils/find.go:134]: The f string parameter might be named "path" or something that would help reader understand the meaning.

File: rules/internal/utils/extension.go [hunk 4]
```diff
@@ -147,3 +147,22 @@ func GetResourceReference(f *desc.FieldDescriptor) *apb.ResourceReference {
 	}
 	return nil
 }
+
+// FindResource attempts to find the google.api.resource(_definition) given the
```
Comment (GregFurmanek) [inline rules/internal/utils/extension.go:151]: Returns first resource of type matching the reference param.

File: rules/internal/utils/extension_test.go [hunk 5]
```diff
@@ -312,3 +312,56 @@ func TestGetResourceReference(t *testing.T) {
...
+	})
+
+	for _, tst := range []struct {
+		name, reference string
+	}{
+		{"local_reference", "library.googleapis.com/Shelf"},
+		{"imported_reference", "library.googleapis.com/Book"},
+	} {
+		t.Run(tst.name, func(t *testing.T) {
+			if got := FindResource(tst.reference, files["shelf.proto"]); got == nil {
```
Comment (GregFurmanek) [inline rules/internal/utils/extension_test.go:360]: Can this be on a separate line?

In a way this single line notation hides the proper usage of the method with method's result verification.

let result = FindResource(...)

if result == nil {
 // error here
 return
}

if resutl.GetType() != tst.reference {
 // A different type of error
}

File: rules/internal/utils/extension_test.go [hunk 6]
```diff
@@ -312,3 +312,56 @@ func TestGetResourceReference(t *testing.T) {
 		}
 	})
 }
+
+func TestFindResource(t *testing.T) {
```
Comment (GregFurmanek) [inline rules/internal/utils/extension_test.go:316]: Should there be a test that doesn't find a resource with requested type?

File: rules/internal/utils/http_test.go [hunk 7]
```diff
@@ -148,3 +148,33 @@ func TestGetVariables(t *testing.T) {
 		})
 	}
 }
+
+func TestHasHTTPRules(t *testing.T) {
+	for _, tst := range []struct {
+		name       string
+		Annotation string
+	}{
+		{"has_rule", `option (google.api.http) = {get: "/v1/foos"};`},
```
[the remark this answers is not in the document]
Comment (noahdietz) [reply rules/internal/utils/http_test.go:157]: The properties of the `google.api.http` method do not affect the ability of the library to extract it from the MethodOptions descriptor, so testing different values there will not have any effect on it.

File: rules/internal/utils/find_test.go (hunk 2 above)
Comment (noahdietz) [reply rules/internal/utils/find_test.go]: Done.

File: rules/internal/utils/find.go (hunk 3 above)
Comment (noahdietz) [reply rules/internal/utils/find.go]: I've already got a variable named `path` in the method body, so I've renamed `f` to `ref`.

File: rules/internal/utils/extension_test.go (hunk 5 above)
Comment (noahdietz) [reply rules/internal/utils/extension_test.go]: Done.

File: rules/internal/utils/extension.go (hunk 4 above)
Comment (noahdietz) [reply rules/internal/utils/extension.go]: Done.

File: rules/internal/utils/http_test.go [hunk 8]
```diff
@@ -148,3 +148,33 @@ func TestGetVariables(t *testing.T) {
...
+						{{ .Annotation }}
+					}
+				}
+
+				message ListFoosRequest {}
+				message ListFoosResponse {}
+			`, tst)
+
+			want := tst.Annotation != ""
+			if got := HasHTTPRules(file.GetServices()[0].GetMethods()[0]); got != want {
```
Comment (GregFurmanek) [inline rules/internal/utils/http_test.go:175]: This could also be split between the execution and check.

Comment (noahdietz) [reply rules/internal/utils/http_test.go]: Done.

File: rules/internal/utils/find_test.go [hunk 9]
```diff
@@ -43,3 +44,46 @@ func TestFindMessage(t *testing.T) {
...
+		name, path string
+	}{
+		{"top_level", "parent"},
+		{"nested", "book.name"},
+		{"double_nested", "book.publishing_info.publisher"},
+	} {
+		t.Run(tst.name, func(t *testing.T) {
+			split := strings.Split(tst.path, ".")
+			want := split[len(split)-1]
+			if f := FindFieldDotNotation(msg, tst.path); f == nil {
```
[the remark this answers is not in the document]
Comment (noahdietz) [reply rules/internal/utils/find_test.go]: Done.

**Review by GregFurmanek — APPROVED** [review]
LGTM

**Pull request merged**